### PR TITLE
Update DevOps tooling

### LINKED ISSUE #550: Update DevOps tooling
`setup.py`

- [x] Update `setup()` contents
- [x] Add `test_requires` which installs dependencies only used in the test suite

Anaconda:

- [x] Remove unused `meta.yaml` (packages are now released through conda-forge feedstock)
  - [x] Remove `meta.yaml` from `tbump.toml`
- [x] Rename `conda/` dir to `conda-env/`
- [x] Rename `e3sm_diags_env_dev.yml` to `dev.yml` and update references
- [x] Rename `e3sm_diags_env.yml` to `prod.yml` and update references 
- [x] Update dependencies in `dev.yml` and `prod.yml` to the latest versions
- [x] Add `ci.yml` for a minimal CI/CD conda env to speed up GH Actions

CI/CD:

- [ ] Add Python matrix testing to `build_workflow.yml`
- [x] Use `ci.yml` instead of `dev.yml` when building conda env

### PR DESCRIPTION
**Pull request by tomvothecoder**
This PR updates the DevOps tooling to optimize build time, remove deprecated files, and update dependencies to the latest versions.

- Closes #550 

## Summary of Changes

Conda environments:

- [x] Rename `conda/` dir to `conda-env/`
- [x] Rename `e3sm_diags_env_dev.yml` to `dev.yml` and `e3sm_diags_env.yml` to `prod.yml`
  - [x]   Update dependencies to the latest versions
  - [x]   Update name and prefix of envs from `e3sm_diags_env_dev` and `e3sm_diags_env_prod` to `e3sm_diags_dev` and `e3sm_diags_prod`
  - [x]   Remove dev dependencies in `prod.yml` (installing latest stable version already includes these dependencies)
- [x] Add `ci.yml` for a minimal CI/CD conda env to speed up GH Actions
  - [x] Only contains the minimum dependencies required to build the package and run the test suite
- [x] Add `docs.yml` for a minimal CI/CD docs env to speed up GH Actions
  - [x] Only contains the minimum dependencies required to build the documentation

  
Conda recipe:

- [x] Remove unused `meta.yaml` 
  - Packages are now released through conda-forge feedstock. 

setup.py:
- [x] Update `setup()` contents with supported Python versions

setup.cfg and pre-commit-config.yaml:
- [x] Update dependencies to the latest versions

CI/CD:

- [x] Add matrix testing for Python 3.8 and 3.9 in `build_workflow.yml`
- [x] Replace pip with conda env (`docs.yml`) in `publish-docs` step

## What to do after this is merged

1. Pull the latest `master`
2. Make sure you are in the root of the repo
3. Recreate environments using yml
  * dev environment - `conda env create -f conda-env/dev.yml`
  * prod environment - `conda env create -f conda-env/prod.yml`
4. `e3sm_diags_dev` and `e3sm_diags_prod` envs are created

### CODE REVIEW

File: .github/workflows/release_workflow.yml [hunk 1]
```diff
@@ -34,7 +34,7 @@ jobs:
             ${{ runner.os }}-
 
       # Using pip for Sphinx dependencies because it takes too long to reproduce a conda environment (~10 secs vs. 3-4 mins)
-      # Low risk trade-off with mismatching dependencies in `e3sm_diags_env_dev.yml`
+      # Low risk trade-off with mismatching dependencies in `e3sm_diags_dev.yml`
```
Comment (tomvothecoder) [inline .github/workflows/release_workflow.yml:37]:
```suggestion
```

File: e3sm_diags/plot/__init__.py [hunk 2]
```diff
@@ -24,8 +24,7 @@ def _get_plot_fcn(backend, set_name):
 
         mod_str = "e3sm_diags.plot.{}.{}_plot".format(backend, set_name)
         module = importlib.import_module(mod_str)
-        # FIXME: error: Module has no attribute "plot"
-        return module.plot  # type: ignore
```
Comment (tomvothecoder) [inline e3sm_diags/plot/__init__.py:28]: This is no longer being caught by `mypy`.

File: e3sm_diags/e3sm_diags_driver.py [hunk 3]
```diff
@@ -267,8 +267,7 @@ def run_diag(parameters):
         mod_str = "e3sm_diags.driver.{}_driver".format(set_name)
         try:
             module = importlib.import_module(mod_str)
-            # FIXME: error: Module has no attribute "run_diag"
-            single_result = module.run_diag(parameters)  # type: ignore
```
Comment (tomvothecoder) [inline e3sm_diags/e3sm_diags_driver.py:271]: This is no longer being caught by the latest version of `mypy`.

File: conda-env/dev.yml [hunk 4]
```diff
@@ -1,47 +1,48 @@
...
   # If versions are updated, also update 'rev' in `.pre-commit.config.yaml`
-  - black=21.5b2
-  - flake8=3.9.2
-  - flake8-isort=4.0.0
-  - mypy=0.812
-  - pre-commit=2.13.0
+  - black=21.12b0
+  - flake8=4.0.1
+  - flake8-isort=4.1.1
+  - isort=5.10.1
```
Comment (tomvothecoder) [inline conda-env/dev.yml:29]: Added `isort` to dependencies

File: setup.py [hunk 5]
```diff
@@ -116,14 +116,26 @@ def get_all_files_in_dir(directory, pattern):
...
 setup(
+    python_requires=">=3.8",
+    classifiers=[
+        "Intended Audience :: Developers",
+        "License :: OSI Approved :: BSD 3-Clause License",
+        "Natural Language :: English",
+        "Programming Language :: Python :: 3",
+        "Programming Language :: Python :: 3.8",
+        "Programming Language :: Python :: 3.9",
+        "Programming Language :: Python :: 3.10",
```
Comment (tomvothecoder) [inline setup.py:127]: Apparently, the package can be built on Python 3.10 based on CI/CD. 

I opted not to update our Anaconda environments to 3.10 just yet though.

File: tests/integration/test_diags.py [hunk 6]
```diff
@@ -82,7 +82,7 @@ def compare_images(
     expected_png = Image.open(path_to_expected_png).convert("RGB")
     diff = ImageChops.difference(actual_png, expected_png)
 
-    diff_dir = "image_check_failures"
+    diff_dir = f"{TEST_ROOT_PATH}image_check_failures"
```
Comment (tomvothecoder) [inline tests/integration/test_diags.py:85]: The `image_check_failures/` dir was being placed in the root of the repo. `.gitignore` wasn't ignoring it as a result.

**Comment by tomvothecoder** [comment]
Hi @forsyth2, I'm hoping to get this PR merged before our next release. If you have the time, it would be great if you can help me fix the failing tests (https://github.com/E3SM-Project/e3sm_diags/runs/4806697230?check_suite_focus=true) and review the PR afterwards.

The changes in this PR should not affect the test suite, which I verified by running  the test suite locally on `master` which also failed.

**Review by forsyth2 — COMMENTED** [review]
Changes look good to me.

As for the test failures, they appear to because of minor plotting changes (probably from the new package versions). Do you want me to update the expected images (https://e3sm-project.github.io/e3sm_diags/_build/html/master/dev_guide/testing.html)?

For `tests/complete_run.py`, it looks like there are many expected images to update as well. Also, `arm_diags/armdiags-convection-onset-twpc3.png` isn't being created for some reason.

File: docs/source/install.rst [hunk 7]
```diff
@@ -133,7 +133,7 @@ an SSL error unless you disable the SSL verification:
 
     ::
 
-        wget https://raw.githubusercontent.com/E3SM-Project/e3sm_diags/master/conda/e3sm_diags_env.yml
+        wget https://raw.githubusercontent.com/E3SM-Project/e3sm_diags/master/conda/prod.yml
```
[the remark this answers is not in the document]
Comment (tomvothecoder) [reply docs/source/install.rst:136]: Thanks, I accidentally reverted that specific update when I force pushed.

The updates are now pushed.

File: docs/source/install.rst [hunk 8]
```diff
@@ -150,7 +150,7 @@ an SSL error unless you disable the SSL verification:
 
     ::
 
-        conda env create -f e3sm_diags_env.yml
+        conda env create -f prod.yml
```
Comment (tomvothecoder) [inline docs/source/install.rst:153]:
```suggestion
        conda env create -f conda-env/prod.yml
```

File: e3sm_diags/derivations/acme.py [hunk 9]
```diff
@@ -159,7 +159,7 @@ def precst(precc, precl):
 
 def tauxy(taux, tauy):
     """tauxy = (taux^2 + tauy^2)sqrt"""
-    var = (taux ** 2 + tauy ** 2) ** 0.5
+    var = (taux**2 + tauy**2) ** 0.5
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:162]: The latest version of `black` (22.1.0) now fixes unnecessary spaces between `**`

**Comment by tomvothecoder** [comment]
Hi @forsyth2 and @chengzhuzhang, this PR is nearly complete, but the test suite is failing (specifically for enso diags).

I validated that the changes in this PR aren't a factor by running the test suite with the latest stable release (v2.6.1) of `e3sm_diags` in a production environment. The expected images might have changed since our last release. 

Please advise on what I should do here, since I usually don't handle test suite failures related to image diffs. Should I update the expected images?

Related log lines:
```
TestAllSets.results_dir=tests/integration/all_sets_results_test/

path_to_actual_png=tests/integration/all_sets_results_test/area_mean_time_series/TREFHT.png
path_to_expected_png=tests/integration//integration_test_images//area_mean_time_series/TREFHT.png
diff has 8 nonzero pixels.
total number of pixels=3825000
num_nonzero_pixels/num_pixels fraction=2.091503267973856e-06

path_to_actual_png=tests/integration/all_sets_results_test/cosp_histogram/MISR-COSP/MISRCOSP-COSP_HISTOGRAM_MISR-ANN-global.png
path_to_expected_png=tests/integration//integration_test_images//cosp_histogram/MISR-COSP/MISRCOSP-COSP_HISTOGRAM_MISR-ANN-global.png
diff has 74 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=3.5175282234105764e-05

path_to_actual_png=tests/integration/all_sets_results_test/enso_diags/TREFHT-response-map/regression-coefficient-trefht-over-nino34.png
path_to_expected_png=tests/integration//integration_test_images//enso_diags/TREFHT-response-map/regression-coefficient-trefht-over-nino34.png
diff has 76581 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=0.036402139037433154

path_to_actual_png=tests/integration/all_sets_results_test/enso_diags/TREFHT-response-map-start-yrs/regression-coefficient-trefht-over-nino34.png
path_to_expected_png=tests/integration//integration_test_images//enso_diags/TREFHT-response-map-start-yrs/regression-coefficient-trefht-over-nino34.png
diff has 76581 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=0.036402139037433154

path_to_actual_png=tests/integration/all_sets_results_test/enso_diags/TREFHT-response-map-test-ref-yrs/regression-coefficient-trefht-over-nino34.png
path_to_expected_png=tests/integration//integration_test_images//enso_diags/TREFHT-response-map-test-ref-yrs/regression-coefficient-trefht-over-nino34.png
diff has 76581 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=0.036402139037433154

path_to_actual_png=tests/integration/all_sets_results_test/enso_diags/TREFHT-response-scatter/feedback-TREFHT-NINO3-TS-NINO3.png
path_to_expected_png=tests/integration//integration_test_images//enso_diags/TREFHT-response-scatter/feedback-TREFHT-NINO3-TS-NINO3.png
diff has 28769 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=0.013675103980986335

path_to_actual_png=tests/integration/all_sets_results_test/enso_diags/TREFHT-response-scatter-start-yrs/feedback-TREFHT-NINO3-TS-NINO3.png
path_to_expected_png=tests/integration//integration_test_images//enso_diags/TREFHT-response-scatter-start-yrs/feedback-TREFHT-NINO3-TS-NINO3.png
diff has 28769 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=0.013675103980986335

path_to_actual_png=tests/integration/all_sets_results_test/enso_diags/TREFHT-response-scatter-test-ref-yrs/feedback-TREFHT-NINO3-TS-NINO3.png
path_to_expected_png=tests/integration//integration_test_images//enso_diags/TREFHT-response-scatter-test-ref-yrs/feedback-TREFHT-NINO3-TS-NINO3.png
diff has 28769 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=0.013675103980986335

path_to_actual_png=tests/integration/all_sets_results_test/lat_lon/ERA-Interim/ERA-Interim-T-850-ANN-global.png
path_to_expected_png=tests/integration//integration_test_images//lat_lon/ERA-Interim/ERA-Interim-T-850-ANN-global.png
diff has 86 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=4.0879382055852646e-05

path_to_actual_png=tests/integration/all_sets_results_test/lat_lon/ERA-Interim/ERA-Interim-T-850-ANN-CONUS_RRM.png
path_to_expected_png=tests/integration//integration_test_images//lat_lon/ERA-Interim/ERA-Interim-T-850-ANN-CONUS_RRM.png
diff has 170 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=8.080808080808081e-05

path_to_actual_png=tests/integration/all_sets_results_test/meridional_mean_2d/ERA-Interim/ERA-Interim-T-ANN-global.png
path_to_expected_png=tests/integration//integration_test_images//meridional_mean_2d/ERA-Interim/ERA-Interim-T-ANN-global.png
diff has 74 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=3.5175282234105764e-05

path_to_actual_png=tests/integration/all_sets_results_test/polar/ERA-Interim/ERA-Interim-T-850-ANN-polar_S.png
path_to_expected_png=tests/integration//integration_test_images//polar/ERA-Interim/ERA-Interim-T-850-ANN-polar_S.png
diff has 532 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=0.0002528817587641117
tests/integration/all_sets_results_test/qbo/qbo-test/qbo_diags.png

path_to_actual_png=tests/integration/all_sets_results_test/qbo/qbo-test/qbo_diags.png
path_to_expected_png=tests/integration//integration_test_images//qbo/qbo-test/qbo_diags.png
diff has 30 nonzero pixels.
total number of pixels=1960000
num_nonzero_pixels/num_pixels fraction=1.530612244897959e-05

path_to_actual_png=tests/integration/all_sets_results_test/streamflow/RIVER_DISCHARGE_OVER_LAND_LIQ_GSIM/seasonality_map.png
path_to_expected_png=tests/integration//integration_test_images//streamflow/RIVER_DISCHARGE_OVER_LAND_LIQ_GSIM/seasonality_map.png
diff has 6 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=2.8520499108734404e-06

path_to_actual_png=tests/integration/all_sets_results_test/streamflow/RIVER_DISCHARGE_OVER_LAND_LIQ_GSIM/annual_map.png
path_to_expected_png=tests/integration//integration_test_images//streamflow/RIVER_DISCHARGE_OVER_LAND_LIQ_GSIM/annual_map.png
diff has 58 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=2.756981580510992e-05

path_to_actual_png=tests/integration/all_sets_results_test/zonal_mean_2d/ERA-Interim/ERA-Interim-T-ANN-global.png
..FFFFFF....F.....
======================================================================
FAIL: test_enso_diags_map (tests.integration.test_diags.TestAllSets)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 372, in test_enso_diags_map
    self.check_enso_map_plots(case_id)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 278, in check_enso_map_plots
    self.check_html_image(html_path, png_path, full_png_path)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 199, in check_html_image
    self.assertEqual(mismatched_images, [])
AssertionError: Lists differ: ['regression-coefficient-trefht-over-nino34.png'] != []

First list contains 1 additional elements.
First extra element 0:
'regression-coefficient-trefht-over-nino34.png'

- ['regression-coefficient-trefht-over-nino34.png']
+ []

======================================================================
FAIL: test_enso_diags_map_start_yrs (tests.integration.test_diags.TestAllSets)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 376, in test_enso_diags_map_start_yrs
    self.check_enso_map_plots(case_id)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 278, in check_enso_map_plots
    self.check_html_image(html_path, png_path, full_png_path)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 199, in check_html_image
    self.assertEqual(mismatched_images, [])
AssertionError: Lists differ: ['regression-coefficient-trefht-over-nino34.png'] != []

First list contains 1 additional elements.
First extra element 0:
'regression-coefficient-trefht-over-nino34.png'

- ['regression-coefficient-trefht-over-nino34.png']
+ []

======================================================================
FAIL: test_enso_diags_map_test_ref_yrs (tests.integration.test_diags.TestAllSets)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 380, in test_enso_diags_map_test_ref_yrs
    self.check_enso_map_plots(case_id)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 278, in check_enso_map_plots
    self.check_html_image(html_path, png_path, full_png_path)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 199, in check_html_image
    self.assertEqual(mismatched_images, [])
AssertionError: Lists differ: ['regression-coefficient-trefht-over-nino34.png'] != []

First list contains 1 additional elements.
First extra element 0:
'regression-coefficient-trefht-over-nino34.png'

- ['regression-coefficient-trefht-over-nino34.png']
+ []

======================================================================
FAIL: test_enso_diags_scatter (tests.integration.test_diags.TestAllSets)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 384, in test_enso_diags_scatter
    self.check_enso_scatter_plots(case_id)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 294, in check_enso_scatter_plots
    self.check_html_image(html_path, png_path, full_png_path)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 199, in check_html_image
    self.assertEqual(mismatched_images, [])
AssertionError: Lists differ: ['feedback-TREFHT-NINO3-TS-NINO3.png'] != []

First list contains 1 additional elements.
First extra element 0:
'feedback-TREFHT-NINO3-TS-NINO3.png'

- ['feedback-TREFHT-NINO3-TS-NINO3.png']
+ []

======================================================================
FAIL: test_enso_diags_scatter_start_yrs (tests.integration.test_diags.TestAllSets)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 388, in test_enso_diags_scatter_start_yrs
    self.check_enso_scatter_plots(case_id)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 294, in check_enso_scatter_plots
    self.check_html_image(html_path, png_path, full_png_path)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 199, in check_html_image
    self.assertEqual(mismatched_images, [])
AssertionError: Lists differ: ['feedback-TREFHT-NINO3-TS-NINO3.png'] != []

First list contains 1 additional elements.
First extra element 0:
'feedback-TREFHT-NINO3-TS-NINO3.png'

- ['feedback-TREFHT-NINO3-TS-NINO3.png']
+ []

======================================================================
FAIL: test_enso_diags_scatter_test_ref_yrs (tests.integration.test_diags.TestAllSets)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 392, in test_enso_diags_scatter_test_ref_yrs
    self.check_enso_scatter_plots(case_id)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 294, in check_enso_scatter_plots
    self.check_html_image(html_path, png_path, full_png_path)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 199, in check_html_image
    self.assertEqual(mismatched_images, [])
AssertionError: Lists differ: ['feedback-TREFHT-NINO3-TS-NINO3.png'] != []

First list contains 1 additional elements.
First extra element 0:
'feedback-TREFHT-NINO3-TS-NINO3.png'

- ['feedback-TREFHT-NINO3-TS-NINO3.png']
+ []

======================================================================
FAIL: test_polar (tests.integration.test_diags.TestAllSets)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 404, in test_polar
    self.check_plots_plevs("polar", "polar_S", [850.0])
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 254, in check_plots_plevs
    self.check_plots_generic(
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 241, in check_plots_generic
    self.check_html_image(html_path, png_path, full_png_path)
  File "/home/runner/work/e3sm_diags/e3sm_diags/tests/integration/test_diags.py", line 199, in check_html_image
    self.assertEqual(mismatched_images, [])
AssertionError: Lists differ: ['ERA-Interim-T-850-ANN-polar_S.png'] != []

First list contains 1 additional elements.
First extra element 0:
'ERA-Interim-T-850-ANN-polar_S.png'

- ['ERA-Interim-T-850-ANN-polar_S.png']
+ []

----------------------------------------------------------------------
Ran 18 tests in 188.050s

FAILED (failures=7)
path_to_expected_png=tests/integration//integration_test_images//zonal_mean_2d/ERA-Interim/ERA-Interim-T-ANN-global.png
diff has 126 nonzero pixels.
total number of pixels=2103750
num_nonzero_pixels/num_pixels fraction=5.9893048128342244e-05
```

**Comment by forsyth2** [comment]
@tomvothecoder 

> The expected images might have changed since our last release.
That's almost certainly what happened

> Please advise on what I should do here, since I usually don't handle test suite failures related to image diffs. Should I update the expected images?
Let me investigate. I'll let you know

**Comment by forsyth2** [comment]
@tomvothecoder Well I have tests passing on `master`, but I can't appear to build the environment being used in this PR to test it:
```
$ cd /home/ac.forsyth2/e3sm_diags
$ git checkout master
$ git fetch upstream master
$ git rebase upstream/master

# Steps from https://e3sm-project.github.io/e3sm_diags/_build/html/master/install.html:
$ conda clean --all
$ conda env create -f conda/e3sm_diags_env_dev.yml -n diags_dev_v261
$ conda activate diags_dev_v261

$ ./tests/clean_tests_repo.sh

# Follow steps from https://e3sm-project.github.io/e3sm_diags/_build/html/master/dev_guide/testing.html:
$ pip install .
$ ./tests/test.sh
# All tests pass

$ git fetch upstream
$ git checkout -b test-pr-551 upstream/devops/550-ci-matrix
$ conda deactivate

# Steps from https://e3sm-project.github.io/e3sm_diags/_build/html/master/install.html:
$ conda clean --all
$ conda env create -f conda-env/dev.yml -n diags_dev_pr551
$ conda activate diags_dev_pr551
Collecting package metadata (repodata.json): done
Solving environment: \
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed                                                                     /                                                                    |
Solving environment: /
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed                                                                     |          /                                                         \

UnsatisfiableError: The following specifications were found to be incompatible with each other:
[...]
The following specifications were found to be incompatible with your system:

  - feature:/linux-64::__glibc==2.28=0
  - feature:|@/linux-64::__glibc==2.28=0
  - cdms2=3.1.5 -> libgfortran-ng -> __glibc[version='>=2.17']
  - cdtime=3.1.4 -> libgfortran-ng -> __glibc[version='>=2.17']
  - genutil=8.2.1 -> libgcc-ng[version='>=7.5.0'] -> __glibc[version='>=2.17']
  - lxml=4.7.1 -> libgcc-ng[version='>=7.5.0'] -> __glibc[version='>=2.17']
  - netcdf4=1.5.8 -> openssh -> __glibc[version='>=2.17,<3.0.a0']
  - scipy=1.8.0 -> libgfortran-ng -> __glibc[version='>=2.17']

Your installed version is: 2.28

Note that strict channel priority may have removed packages required for satisfiability.
```

**Comment by forsyth2** [comment]
> I would create a clean environment using `dev.yml`

I thought that's what I was doing with `conda env create -f conda-env/dev.yml -n diags_dev_pr551`. Is that incorrect?

**Comment by xylar** [comment]
I have a tool I use when `mamba` fails to solve packages:
https://gist.github.com/xylar/e02438b5a19e8aa73439ef2bc42b9aea
Replace the list of packages with yours and it will find the first package in the list that causes the solve to fail.  You can re-order the list by putting the failing package at the top to see what other package it conflicts with.

You could modify the tool to use `conda` instead of `mamba` if you don't mind it taking all day...

**Comment by tomvothecoder** [comment]
Thanks for providing that tool @xylar, it looks really useful.

In my case, I took an alternate approach by building the production env with the latest stable version of `e3sm_diags`. I checked if the base dependencies of the prod env aligned with the base dependencies of the dev env, which they did. Through process of elimination, I found the dev env package conflict was related to `sphinx=4.4.0` and some other package(s). I pinned `sphinx=4.3.2` instead and now the dev env builds properly.

I still need to figure out the root cause of the conflict, but at least things are working again for now.

@forsyth2, you can try building the dev env using `dev.yml` and running the test now.

**Comment by forsyth2** [comment]
> you can try building the dev env using dev.yml and running the test now.

@tomvothecoder Ok, I got 7 test failures that appear to be caused by the 3 image failures at https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/e3sm_diags_development/image_check_failures_20220224/. The only noteworthy diff I see is that `feedback-TREFHT-NINO3-TS-NINO3` changes the ordering of its legend. 

So, I think it's ok to just update the expected images as described in https://e3sm-project.github.io/e3sm_diags/_build/html/master/dev_guide/testing.html. Do you want me to do that?

**Comment by tomvothecoder** [comment]
Thanks for checking @forsyth2! That would be awesome if you can handle updating the expected images since I'll be out tomorrow.

**Comment by forsyth2** [comment]
@tomvothecoder Ok the build workflows are passing now. For reference, here are the steps I ran:
```
$ cd /home/ac.forsyth2/e3sm_diags
$ conda activate diags_dev_pr551_20220224
$ pip install .
$ ./tests/clean_tests_repo.sh
$ ./tests/test.sh

# Follow steps from https://e3sm-project.github.io/e3sm_diags/_build/html/master/dev_guide/testing.html
$ cd /lcrc/group/e3sm/public_html/e3sm_diags_test_data/integration/expected
$ cat README.md
E3SM Diags version: v2_5_0
Date these files were generated: 20210706
Hash of commit (first 7 characters): 69b01d2
$ mv integration_test_images previous_output/integration_test_images_v2_5_0_20210706_69b01d2
$ cd /home/ac.forsyth2/e3sm_diags
# TODO: Directions need to be updated
$ mv tests/integration/all_sets_results_test /lcrc/group/e3sm/public_html/e3sm_diags_test_data/integration/expected/integration_test_images
$ cd /lcrc/group/e3sm/public_html/e3sm_diags_test_data/integration/expected

$ ls
integration_test_images  previous_output  README.md  README.md~
$ cd /home/ac.forsyth2/e3sm_diags
$ ./tests/clean_tests_repo.sh
$ ./tests/test.sh
# All tests pass

# TODO: After merging this pull request
# Edit README.md:
E3SM Diags version: v2_6_1
Date these files were generated: 2022_02_25
Hash of commit (first 7 characters): <hash after merging PR>
```
To do:
- [ ] #564 (before next release)
- [ ] Edit `README.md` after this pull request merges

**Comment by tomvothecoder** [comment]
@forsyth2 Great, thanks for taking care of that Ryan. Happy to see the build is passing again!

**Comment by tomvothecoder** [comment]
FYI @forsyth2 and @chengzhuzhang, you should follow the "What to do after this is merged" instructions in this PR.

**Comment by chengzhuzhang** [comment]
Thank you for the heads-up!

**Comment by forsyth2** [comment]
@tomvothecoder Thanks, I updated my environments and also did the update-README step I listed above.

**Pull request merged**